xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,291 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.XdsLoadBalancerProvider.XDS_POLICY_NAME;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Attributes;
+import io.grpc.ChannelLogger;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.NameResolver.ConfigOrError;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.ServiceConfigUtil.LbConfig;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.CdsLoadBalancerProvider.CdsConfig;
+import io.grpc.xds.XdsClient.ClusterUpdate;
+import io.grpc.xds.XdsClient.ClusterWatcher;
+import io.grpc.xds.XdsLoadBalancerProvider.XdsConfig;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_cds LB policy.
+ */
+public final class CdsLoadBalancer extends LoadBalancer {
+  private final ChannelLogger channelLogger;
+  private final LoadBalancerRegistry lbRegistry;
+  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
+  private final Helper helper;
+
+  // The following fields become non-null once handleResolvedAddresses() successfully.
+
+  // Most recent CdsConfig.
+  @Nullable
+  private CdsConfig cdsConfig;
+  // Most recent ClusterWatcher.
+  @Nullable
+  private ClusterWatcher clusterWatcher;
+  @Nullable
+  private ObjectPool<XdsClient> xdsClientRef;
+  @Nullable
+  private XdsClient xdsClient;
+
+  CdsLoadBalancer(Helper helper) {
+    this(helper, LoadBalancerRegistry.getDefaultRegistry());
+  }
+
+  @VisibleForTesting
+  CdsLoadBalancer(Helper helper, LoadBalancerRegistry lbRegistry) {
+    this.helper = helper;
+    this.channelLogger = helper.getChannelLogger();
+    this.lbRegistry = lbRegistry;
+    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(helper);
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+    Attributes attributes = resolvedAddresses.getAttributes();
+    if (xdsClientRef == null) {
+      xdsClientRef = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_REF);
+      if (xdsClientRef == null) {
+        // TODO(zdapeng): create a new xdsClient from bootstrap if no one exists.
```

Comment:
Do we really have such a usage?

Reply:
Currently we don't. But `CdsLoadBalancer` should not assume the name resolver from which it receives addresses from could only be `XdsNameResolver`.